src/onegov/org/models/search.py

```diff
@@ -159,5 +166,221 @@ def subset_count(self) -> int:
+                    and search_result.latest_occurrence
+                    and search_result.latest_occurrence.start > utcnow()):
+                future_events.append(search_result)
+            else:
+                other.append(search_result)
+
+        if not future_events:
+            return batch
+
+        sorted_events = sorted(
+            future_events, key=lambda e:
+            e.latest_occurrence.start,  # type:ignore[attr-defined]
+            reverse=True)
+
+        return sorted_events + other
+
+    def _create_weighted_vector(
+        self,
+        model: Any,
+        language: str = 'simple'
+    ) -> Any:
+        # for now weight the first field with 'A', the rest with 'B'
+        weighted_vectors = [
+            func.setweight(
+                func.to_tsvector(
+                    language,
+                    getattr(model.fts_idx_data, field, '')),
+                weight
+            )
+            for field, weight in zip(model.es_properties.keys(), 'ABBBBBBBBBB')
+            if not field.startswith('es_')  # TODO: rename to fts_
+        ]
+
+        # combine all weighted vectors
+        if weighted_vectors:
+            combined_vector = weighted_vectors[0]
+            for vector in weighted_vectors[1:]:
+                combined_vector = combined_vector.op('||')(vector)
+        else:
+            combined_vector = func.to_tsvector(language, '')
+
+        return combined_vector
+
+    def generic_search(self) -> list['Searchable']:
+        doc_count = 0
+        results: list[Any] = []
+        language = locale_mapping(self.request.locale or 'de_CH')
+        ts_query = func.websearch_to_tsquery(language,
+                                             func.unaccent(self.web_search))
+        session = self.request.session
+
+        for base in self.request.app.session_manager.bases:
+            for model in searchable_sqlalchemy_models(base):
+                query = session.query(model)
+                if session.query(query.exists()).scalar():
+                    weighted = (
+                        self._create_weighted_vector(model, language))
+                    rank_expression = func.coalesce(
+                        func.ts_rank(
+                            weighted,
+                            ts_query,
+                            0  # normalization, ignore document length
+                        ), 0).label('rank')
+                    query = (query.filter(model.fts_idx.op('@@')(ts_query))
+                             .add_columns(rank_expression))
+                    res = list(query.all())
+                    doc_count += len(res)
+                    results.extend(res)
+
+        # exclude non-public documents if user is not logged in
+        # we do this on python level as using hybrid properties and sql
+        # expressions did not work as expected
+        if not self.request.is_logged_in:
+            results = [r for r in results if r[0].es_public == True]
```

Comment:
Test for `public` documents done in python as the publication needs to be tested according the current time and therefore cannot base on static data in column `fts_idx_data`

Reply:
Or as you, @Daverball, suggested earlier we run re-indexing from recently published documents as a cron job...

Reply:
This is already happening for elasticsearch, so it would be more or less free to extend this with a postgres reindex: https://github.com/OneGov/onegov-cloud/blob/master/src/onegov/org/cronjobs.py#L91